src/test/shared/utilities/workspaceUtils.test.ts

```diff
@@ -0,0 +1,82 @@
+ */
+
+import * as assert from 'assert'
+import * as path from 'path'
+import * as sinon from 'sinon'
+import * as vscode from 'vscode'
+import { findParentProjectFile } from '../../../shared/utilities/workspaceUtils'
+
+describe('findParentProjectFile', async () => {
+    const workspaceDir = path.join('code', 'workspace')
+    before(() => {
+        const sandbox = sinon.createSandbox()
+        sandbox.stub(vscode.workspace, 'getWorkspaceFolder').returns({
+            name: 'tempFolder',
+            index: 0,
+            uri: vscode.Uri.parse(workspaceDir),
+        })
+    })
+    const sourceCodeUri = vscode.Uri.file(path.join(workspaceDir, 'someproject', 'src', 'Program.cs'))
+    const projectInSameFolderUri = vscode.Uri.file(path.join(workspaceDir, 'someproject', 'src', 'App.csproj'))
+    const projectInParentFolderUri = vscode.Uri.file(path.join(workspaceDir, 'someproject', 'App.csproj'))
+    const projectInParentParentFolderUri = vscode.Uri.file(path.join(workspaceDir, 'App.csproj'))
+    const projectInParentButOutOfWorkspace = vscode.Uri.file(path.join('code', 'App.csproj'))
+    const projectOutOfParentChainUri = vscode.Uri.file(path.join('code', 'someotherproject', 'App.csproj'))
+
+    const testScenarios = [
+        {
+            scenario: 'locates project in same folder',
+            findFilesResult: [projectInSameFolderUri],
+            expectedResult: projectInSameFolderUri,
+        },
+        {
+            scenario: 'locates project in parent folder',
+            findFilesResult: [projectInParentFolderUri],
+            expectedResult: projectInParentFolderUri,
+        },
+        {
+            scenario: 'locates project two parent folders up',
+            findFilesResult: [projectInParentParentFolderUri],
+            expectedResult: projectInParentParentFolderUri,
+        },
+        {
+            scenario: 'selects project in same folder over parent folder',
+            findFilesResult: [projectInSameFolderUri, projectInParentFolderUri],
+            expectedResult: projectInSameFolderUri,
+        },
+        {
+            scenario: 'always selects project in same folder over parent folder regardless of order',
+            findFilesResult: [projectInParentFolderUri, projectInSameFolderUri],
+            expectedResult: projectInSameFolderUri,
+        },
+        {
+            scenario: 'returns undefined when no project files are located',
+            findFilesResult: [],
+            expectedResult: undefined,
+        },
+        {
+            scenario: 'returns undefined when no project files are located in parent chain',
+            findFilesResult: [projectOutOfParentChainUri],
+            expectedResult: undefined,
+        },
+        {
+            scenario: 'returns undefined when the project file is in the parent chain but out of the workspace folder',
+            findFilesResult: [projectInParentButOutOfWorkspace],
+            expectedResult: undefined,
+        },
```

Comment:
FYI: I may remove the failing test. That test is the only one that tests the new functionality, but the way the function is constructed currently fundamentally breaks that test. Specifically: the value being fed in should be filtered out, but since we're overriding the filter, the logic is sidestepped.

The only way to fix this one is via integration test. I'll see if I can move that there.